### PR TITLE
Kylepressel/rain self collection

### PR DESCRIPTION
**Pull request by pressel**
Port of rain_self_collection subroutine from F90 to C++. This PR is non-B4B due to the use of B4B macros in micro_p3.F90. The non-B4B differences in regression tests are small. The new B4B unit tests exercise all if branches.

### CODE REVIEW

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 1]
```diff
@@ -2787,6 +2791,16 @@ subroutine rain_self_collection(rho,qr_incld,nr_incld,    &
 
    real(rtype) :: dum, dum1, dum2
 
+#ifdef SCREAM_CONFIG_IS_CMAKE
+
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3.F90]: Don't need this newline

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 2]
```diff
@@ -2787,6 +2791,16 @@ subroutine rain_self_collection(rho,qr_incld,nr_incld,    &
...
+   if (use_cxx) then
+      call  rain_self_collection_f(rho,qr_incld,nr_incld,    &
+         nrslf)
+      return
+   endif
+#endif
+
+   !  print *, rho, qr_incld, nr_incld, nrslf
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3.F90]: Remove commented-out print.

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 3]
```diff
@@ -2797,20 +2811,19 @@ subroutine rain_self_collection(rho,qr_incld,nr_incld,    &
...
+         dum = 2._rtype-bfb_exp(2300._rtype*(dum2-dum1))
       endif
 
       if (iparam.eq.1) then
          nrslf = dum*kr*1.e-3_rtype*qr_incld*nr_incld*rho
       elseif (iparam.eq.2 .or. iparam.eq.3) then
          nrslf = dum*5.78_rtype*nr_incld*qr_incld*rho
       endif
-
+
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3.F90]: Remove trailing whitespace. You may need to add trailing-ws highlighting to your editor.

File: components/scream/src/physics/p3/p3_functions_f90.hpp [hunk 4]
```diff
@@ -233,6 +233,26 @@ extern "C"{
 
 ///////////////////////////////////////////////////////////////////////////////
 
+
+struct RainSelfCollectionData
+{
+  //inputs 
+  Real rho;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_f90.hpp]: Combine these.

File: components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+namespace scream {
+namespace p3 {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::rain_self_collection(const Spack& rho, const Spack& qr_incld, const Spack& nr_incld, Spack& nrslf)
+{
+
+// ------------------------------------------------------
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: This comment should be indented.

File: components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// self-collection and breakup of rain
+// (breakup following modified Verlinde and Cotton scheme)
+
+  constexpr Scalar qsmall       = C::QSMALL;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: = should be aligned for these three lines.

File: components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  constexpr Scalar rhow = C::RhoH2O;
+  constexpr Scalar pi = C::Pi;
+
+  const auto qr_incld_not_small = qr_incld >= qsmall; 
+  
+  if(qr_incld_not_small.any()){
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: We typically put a space between `if` and `(`

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+  static void run_rain_self_collection_bfb_tests(){
+
+    using KTH = KokkosTypes<HostDevice>;
+
+    static constexpr Int max_pack_size = 16;
+    REQUIRE(Spack::n <= max_pack_size);
+    
+    RainSelfCollectionData dc[max_pack_size] = {
+    //  rho, qr_incld, nr_incld, nrslf
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: Contents of array should have extra indent.

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    {1.120E+00, 5.650E-05, 8.277E+05, 0.000E+00},
+
+    {1.060E+00, 1.354E-03, 1.401E+04, 0.000E+00},
+    {1.136E+00, 1.319E-04, 4.210E+03, 0.000E+00},
+    {1.166E+00, 5.339E-03, 1.804E+04, 0.000E+00},
+    {1.120E+00, 5.650E-05, 8.277E+05, 0.000E+00}
+    };
+
+    //Sync to device
+	KTH::view_1d<RainSelfCollectionData> dc_device("dc", Spack::n);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: More indenting issues.

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	auto dc_host = Kokkos::create_mirror_view(dc_device);
+
+    std::copy(&dc[0], &dc[0] + Spack::n, dc_host.data());
+    Kokkos::deep_copy(dc_device, dc_host);
+
+	//Get data from fortran 	
+  	for (Int i = 0; i < Spack::n; ++i) {
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: strange indent.

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      rain_self_collection(dc[i]);
+  	}
+
+	//Run function from a kernal and copy results back to the host
+  	Kokkos::parallel_for(RangePolicy(0, 1), KOKKOS_LAMBDA(const Int& i) {
+	  // Init pack inputs
+      Spack rho_local, qr_incld_local, nr_incld_local, nrslf_local;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: indent.

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+          dc_device(s).nrslf = nrslf_local[s];
+	  }
+    });
+
+    // Sync back to host
+    Kokkos::deep_copy(dc_host, dc_device);
+
+    // Validate results
+    for (Int s = 0; s < Spack::n; ++s) {
+        REQUIRE(dc[s].rho == dc_host(s).rho);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: too much indent.

File: components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+}; //TestRainselfCollection
+
+
+} // namespace unit_test
+} // p3
+} // scream
+
+namespace{
+
+    TEST_CASE("p3_rain_self_collection_test", "[p3_rain_self_collection_test"){
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_rain_self_collection_tests.cpp]: strange indentation

File: components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  constexpr Scalar pi     = C::Pi;
+
+  const auto qr_incld_not_small = qr_incld >= qsmall; 
+  
+  if (qr_incld_not_small.any()){
+    const Real dum1 = sp(280.e-6); 
+    const auto dum2 = pack::cbrt((qr_incld)/(pi*rhow*nr_incld));
+    const auto dum2_lt_dum1 = dum2 < dum1;
+    if (dum2_lt_dum1.any()){
+      nrslf.set(dum2_lt_dum1, sp(5.78)*nr_incld*qr_incld*rho);//sp(5.78)*nr_incld*qr_incld*rho);
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: To me it looks like you translated the same second block twice and missed the first block, namely this one:
```if (dum2.lt.dum1) then
         dum = 1._rtype
      else if (dum2.ge.dum1) then
         dum = 2._rtype-exp(2300._rtype*(dum2-dum1))
      endif
```

Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: In the spirit of vectorization, you could compute  `sp(5.78)*nr_incld*qr_incld*rho` just once, save it to a Spack, then use it in both branches.

Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: All of your `set`s need to mask against `qr_incld_not_small` as well as the `dum2 < dum1` inequality.

Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_rain_self_collection_impl.hpp]: If your unit test is not sensitive to this mis-masked set, then please make it so. Or perhaps this is why the unit test is failing. In that case, good unit test!

**Comment by jgfouca** [comment]
@pressel , is this PR ready?

**Comment by jgfouca** [comment]
Tested manually.

**Pull request merged**